mcos/covariance_transformer.py

```diff
@@ -177,24 +207,48 @@ def _de_noised_corr(self, eigenvalues: np.array, eigenvectors: np.array, n_facts
         corr = cov_to_corr(corr)
         return corr
 
-    def _corr_to_cov(self, corr: np.array, std: np.array) -> np.array:
+
+class DetoneCovarianceTransformer(AbstractCovarianceTransformer):
+    def __init__(self, n_remove: int):
         """
-        Recovers the covariance matrix from the de-noise correlation matrix
-        :param corr: de-noised correlation matrix
-        :param std: standard deviation of the correlation matrix
-        :return: a recovered covariance matrix
+        Removes the largest eigenvalue/eigenvector pairs from the covariance matrix. Since the largest eigenvalues are
+        typically associated with the market component, removing such eigenvalues has the effect of removing the
+        market's influence on the correlations between securities. See chapter 2.6 of "Machine Learning for Asset
+        Managers".
+        :param n_remove: The number of the largest eigenvalues to remove
         """
-        cov = corr * np.outer(std, std)
-        return cov
+        self.n_remove = n_remove
 
+    def transform(self, cov: np.array, n_observations: int) -> np.array:
+        if self.n_remove == 0:
+            return cov
 
-def cov_to_corr(cov: np.array) -> np.array:
-    """
-    Derive the correlation matrix from a covariance matrix
-    :param cov: covariance matrix
-    :return: correlation matrix
-    """
-    std = np.sqrt(np.diag(cov))
-    corr = cov / np.outer(std, std)
-    corr[corr < -1], corr[corr > 1] = -1, 1  # numerical error
-    return corr
+        corr = cov_to_corr(cov)
+
+        w, v = linalg.eig(corr)
+
+        # sort from highest eigenvalues to lowest
+        sort_index = np.argsort(-np.abs(w))  # get sort_index in descending absolute order - i.e. from most significant
+        w = w[sort_index]
+        v = reorder_matrix(v, sort_index)
```

Comment:
reorder_matrix reorders both columns and rows of the matrix. won't reordering the rows of the eigenvectors change the vectors themselves?